File: PYTHON/iterator/iterator.py
class iterador:

    def __init__(self, numberElements, size):
        self.numberElements = numberElements
        self.size = size
        self.reset()

    def GetComb(self):
        result = self.comb.copy()
        return result

    def newComb(self):
        last = self.size - 1
        if self.comb [last] < (self.numberElements - self.size + last):
            self.comb[last] = self.comb[last] + 1
        else:

            j = last
            while self.comb[j] == (self.numberElements - self.size + j ):
                if j == 0:
                    return False
                j = j - 1
            self.comb[j] = self.comb[j] + 1   
            for k in range(j + 1,self.size):
                self.comb[k] = self.comb[k-1] + 1
        return True
            
    def hasNext(self):
        return self.newComb()

    def reset(self):
        self.comb = []
        for i in range(0,self.size):
            self.comb.append(i)
        
        self.comb[self.size - 1] = self.comb[self.size - 1] - 1 

File: GO/iterator/iterator.go
package iterator

// Iterator - for iterating through combinations
// combination []int - Last combination array
// numberElements int - Number of elements to combinate with
// combinationSize int - Size of each combination
type Iterator struct {
	combination                     []int
	numberElements, combinationSize int
}

// InitIterator - Initilization of the iterator
// numberElements int - numberElements to choose from when combinating
// combinationSize int - size of the combinations
func InitIterator(numberElements, combinationSize int) *Iterator {
	iter := Iterator{}
	iter.numberElements = numberElements
	iter.combinationSize = combinationSize
	iter.Reset()
	return &iter
}

// Next - Return a []int representing one combination
func (iter *Iterator) Next() []int {
	result := make([]int, iter.combinationSize)
	copy(result, iter.combination)
	return result
}

// NewComb - Creates a new combination
func (iter *Iterator) newComb() bool {
	last := iter.combinationSize - 1
	if iter.combination[last] != (iter.numberElements - iter.combinationSize + last) {
		iter.combination[last]++
	} else {
		i := last
		for iter.combination[i] == (iter.numberElements - iter.combinationSize + i) {
			if i <= 0 {
				return false
			}
			i--
		}
		iter.combination[i]++
		for j := i + 1; j < iter.combinationSize; j++ {
			iter.combination[j] = iter.combination[j-1] + 1
		}
	}
	return true
}

// HasNext - returns true if there is another combination to get
func (iter *Iterator) HasNext() bool {
	return iter.newComb()
}

// Reset - Resets the iterator
func (iter *Iterator) Reset() {
	for i := 0; i < iter.combinationSize; i++ {
		iter.combination = append(iter.combination, i)
	}
	iter.combination[iter.combinationSize-1]--
}
